### ISSUE TITLE
Support simple config file as data provider (for users definition)

### ISSUE THREAD

**Issue by thomsh** [drakkan/sftpgo#66]
Hi, 
You have started a **very** nice project ^_^
I would like to run SFTPGo in a higly secure environment or at least ask security team for the permission ^^'

I'm looking for a file server (protocol can be sftp, or else) with virtual user / non unix user, with everything in immutable configuration file. 
Config can be generated by a config manager like chef, ansible watever.
You already have conf key to disable web interface etc.
So, I suggest a config file as data provider. 

Have a nice day !

**Comment by drakkan** [comment]
Hi,

thanks for your suggestion,

adding a file based data provider is not too difficult, anyway I think its management it is not so easy, for example the password for the users must be stored as hash and not as clear text, the public keys are very long strings and this will make difficult to have a simple line based users store, we need a structured text file, something like JSON/YAML or such.

An immutable users store can be achieved already, simply set `manage_users` and `track_quota` to 0 in your config file. 

You can enforce a read only data provider using:

- a database view for SQL based providers
- a read only file for SQLite or bolt provider (for example `chmod 444 sftpgo.db` or you can store the database on a read only filesystem)

if you want to add users manually using a script you can do this with the sqlite provider, for example:

```
sqlite3 <path to db> '<INSERT user query>'
```

or even for SQL based data providers for example using the `psql` utility for PostgreSQL and so on.

So basically before adding a text based data provider I would understand what's wrong with the current approach.

Do you know other server software that stor users in a way acceptable for your use case? A such example could be useful to better evaluate your use case.

Additionally a file based data provider should be loaded in memory. 

Maybe I can add an option to populate the current memory provider from a backup file generated using the `dumpdata` REST API, but I would like to better understand your use case/rationale before.

Thanks!

**Comment by thomsh** [comment]
Hi,
Thanks for your quick reply !

### About using sqlite as a pre-provisionned read only config
I tried your suggestion.
To do so, I created the sqlite with your scripts:
`find sql/sqlite/ -type f -iname '*.sql' -print |sort -n|xargs cat |sqlite3 sftpgo.db`

Then I looked on the README, plus a previous sqlitefile filled with `sftpgo` web interface.
Then I deduce the simplest query to add user with all perm on a folder with pub key auth:
```sql
INSERT INTO users 
(username, public_keys, home_dir, uid, gid,max_sessions, quota_size, quota_files, permissions, used_quota_size, used_quota_files, last_quota_update, upload_bandwidth, download_bandwidth, expiration_date, last_login, status) 
values ('toto', 
'["ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIJN3NhVG7tTZvjETwNF42t4mm0FvvFQjkM1xfkJNi5kV"]',
 '/tmp/',
 0,0,0,0,0, '{"/":["*"]}', 0, 0, 0, 0, 0, 0, 0, 1);
```
**Technically It does the job !**

#### My thoughts after this test:
- Sqlite (or similar data provider) have update on the fly which is good to avoid restarting SFTPGo, as you don't support an unix signal to reload the config?
- Filling the database with an external script:
..- Not practical, but still ok
..- Script will Required some maintenance as SFTPGo schema will evolve one day
..- Need to find all the default value for lot of fields, deep look inside SFTPGo required
..- For thoses who use password need to precompute  hash & store it in the right format.
- A little hard to review active config than a yaml file for example.

**But I think filling up the DB yourself or with a script is always the wrong, This is why you have a REST API and python CLI ^_^**

So, if your are not allowed to let that open API running:
Simplest possible solutions:
- Add  auth on the API itself (required dev on SFTPGo). _like gen auth basic, set in the SFTPGo config, and use the python cli_
- Or start the API only for server maintenance: not great.
- Or protect the API with firewall, even from lo, to let only the users you want to call the api.
- Or having a command line option like `sftpgo config add-user -f conf.yaml --name toto --key 'xx' etc` to fill the data provider, and maybe run schema creation too (dev required on SFTPGo), sadly it's close to the  python cli you already have.

The idea is having a file only accessible by root to configure access is OK
Having any unix user (mean, any app), able to connect on the `lo` interface to reconfigure access is NOK

Again, my personal thoughts, for my requirements.

### answers

> adding a file based data provider is not too difficult, anyway I think its management it is not so easy, for example the password for the users must be stored as hash and not as clear text, the public keys are very long strings and this will make difficult to have a simple line based users store, we need a structured text file, something like JSON/YAML or such.

- Yes a structured config file is required. We should config only the minimum, like in your web interface.
- Need to give in the doc a compatible command to generate password. (when password are used). e.g. like `mkpasswd --method=SHA-512 --stdin` for /etc/shadow
- Adding a signal handling to reload conf can help, as restart will cut active upload
- Maybe (or not) use this config file to fillup an actual data provider ?

> So basically before adding a text based data provider I would understand what's wrong with the current approach.

Not really wrong !
Not really practical to provisioned server with config management, in a way everything is immutable, audit-ready, config via an external tool.

> Do you know other server software that stor users in a way acceptable for your use case? A such example could be useful to better evaluate your use case.

Not really, still looking for, but:
openssh: generating config file (sshd_config, authorized_keys) plus run useradd -p '<hash>'
nginx with mod_dav: generating nginx.conf and a htpasswd file for basic auth.
_note in my use case openssh don't really match, main problem: bound to unix users_

Again, nice feature, I'm glad I found this app,
Thanks
Have a nice day

**Comment by jovandeginste** [comment]
I like this suggestion, as it supports "infrastructure as code", and allows for versioning of the (user) configuration.

I agree the configuration has to be a structured format, be it yaml or json (preferably any/every format you already support for the general configuration)

Finally, I think the daemon should reload the configuration "live". Not sure if it should happen automatically or after a signal from the service manager. Viper supports live/automatic config updates (`WatchConfig`), maybe this can be used for this?

**Comment by drakkan** [comment]
Hi, thanks for the clarifications,

I'll try my best to make sftpgo fit this use case too, 

so basically some useful additions:

- password protection for the REST API. This should solve your issue, can you please confirm?
- adding a mkpasswd subcommand could be useful too

A subcommand to fill the data provider is not a viable solution:

- it works with MySQL and PostgreSQL
- it will not work with bolt since bolt lock the database file so you can't use this subcommand while sftpgo is running
- it could cause issues with SQLite, for example occasional "database is locked" errors. 

Regarding the file based data provider it could work but there are some limitations:

1) a modification, for example last login update or quota update, will require rewriting the whole file
2) lookup for a user will require parsing the whole file too

so if I implement a such provider the users will be loaded at startup and after modifications (using viper or a signal or both), basically we will use this file to fill the memory provider (currently used for portable mode only), quota updates and last login will not be persisted. I don't want to reinvent a database :)

I'm open to suggestions, thanks!

P.S. you could be interested to the external authentication too

**Comment by jovandeginste** [comment]
Totally agree, the file should _not_ be modified by the service! Only read it.

**Comment by drakkan** [comment]
If REST API are not disabled you can use `dumpdata` and `loaddata` to achieve something similar too.

For example you prepare your users in a test environment and then export then using `dumpdata`, a JSON file with the users definition will be generated.

This dump can be loaded on your prod environment (even if prod and dev use different data providers), existing users will be updated the new ones will be added. I could also add an option to only add the new users from a dump.

This is not to say that I don't want to support a file based provider, is only to show another option.

**Comment by thomsh** [comment]
Hi,

> so basically some useful additions:
> 
>     * password protection for the REST API. This should solve your issue, can you please confirm?
> 

Not really, but still a good workaround, I think it's acceptable (hope)


> A subcommand to fill the data provider is not a viable solution:
> 
>     * it works with MySQL and PostgreSQL
> 
>     * it will not work with bolt since bolt lock the database file so you can't use this subcommand while sftpgo is running
> 
>     * it could cause issues with SQLite, for example occasional "database is locked" errors.

Ok, thanks, should be avoid then !


> so if I implement a such provider the users will be loaded at startup and after modifications (using viper or a signal or both), basically we will use this file to fill the memory provider (currently used for portable mode only), quota updates and last login will not be persisted. I don't want to reinvent a database :)

yes sure, quota or last login cannot be persisted.

> P.S. you could be interested to the external authentication too

Yes, true, I should dig

#### About dumping & loading data
yep it's true, but don't really work in some infra as code & env design i'm facing. 

So many kind of use case ^_^''

Thanks

**Comment by drakkan** [comment]
Hi,

my plan is to add to the memory provider an option to load the users from a file. I did some tests using viper and it works.

An example file, for example in JSON, is:

```
{"users":[{"username":"user1",.....},{"username":"user2"....},...]
```
viper supports several formats, so theoretically you could write the users file as YAML, TOML ecc... 

Since this is the same format used in dumpdata you could use a private sftpgo instance to create and manage the users, using for example the web interface. From this private instance you take a dump and use this dump as file provider for your production sftpgo instance.

I tested viper hot reload capability but sometimes it does not work as expected, see this bug:

https://github.com/spf13/viper/issues/174

viper use fsnotify so you can also have multiple events too, for example saving my test user files using `vi` generate 3 reload events (probably `vi` does 3 writes), I don't like this. 

I think it is better to have an explicit reload event and/or subcommand. On Linux the obvious solution is a SIGHUP but this will not work on Windows (and you don't want a REST API for this), so I have to think a bit more about this (on windows I have to investigate how to receive this event inside the windows service).

Writing a users file by hand is really hard, you need to know fields name and their format, make the required changes each time a new field is added/deleted ecc... 
Do you really want to write this file by hand? I think it is much better to use a dump from a test instance, do you agree?

If we use the dump and we don't use viper watching feature probably viper is useless too.

**Comment by drakkan** [comment]
I found a way to notify the running Windows service, so my current plan is:

- add initial users to the memory provider from a dump file reusing the current settings, for example:

```
 "data_provider": {
    "driver": "memory",
    "name": "dump.json",
```

where `dump.json` is obtained using `dumpdata` (even from another instance running with a different data provider on a private server) and it will be loaded at startup
- this file can be reloaded on demand sending a sighup on Linux/Unix and a `paramchange` request to the running service on Windows. Changes for existing users will be applied at the next login.

Eventually if we really want to modify the users file by hand, in future we can add support for dumping data to a format such as YAML/TOML that allows to add comments and it is probably more readable than JSON. Anyway in this case we need to add some new subcommands to generate an hashed password and to validate the user generated file. Currently I don't like this solution, I think it is better to use a dump that is already validated and if we add new fields you simply need to get a new dump instead of manually edit this file.

Not strictly related to this issue: I plan to add basic auth and https support for the REST API too

**Comment by thomsh** [comment]
Hi @drakkan,
I agreed with you, viper detecting file change can be hazardous.

> Do you really want to write this file by hand? I think it is much better to use a dump from a test instance, do you agree?

Not from scratch, on my side I'll plan use the dump to generate the config & make sure the format is correct then convert it as a template, to be fill by a config management (chef, ansible, pupet etc)

+1 for the basic auth & https on the API

Thanks !

Have a nice day

**Comment by drakkan** [comment]
> Hi @drakkan,
> I agreed with you, viper detecting file change can be hazardous.
> 
> > Do you really want to write this file by hand? I think it is much better to use a dump from a test instance, do you agree?
> 
> Not from scratch, on my side I'll plan use the dump to generate the config & make sure the format is correct then convert it as a template, to be fill by a config management (chef, ansible, pupet etc)

you should use the indent mode for the dump to make it more readable.

If you think that your template could be useful to others too please send a pull request, thanks!

> 
> +1 for the basic auth & https on the API

it is now supported

https://github.com/drakkan/sftpgo/commit/8b039e0447fa1ddb2e065637db905827d8aa69f2

> 
> Thanks !

Thank you for your suggestions, you helped me to improve sftpgo!
> 
> Have a nice day

**Comment by mcarbonneaux** [comment]
> viper use fsnotify so you can also have multiple events too, for example saving my test user files using `vi` generate 3 reload events (probably `vi` does 3 writes), I don't like this.

i'm facing the same problem in some of my code, i've resolved by using ticker of 1s, that are started/reset on each event, and at the last event the ticker event is receive and you can reload....  this delais the reload 1s after the last fsnotify event ...

but i've been face of another problem in conjonction with gitsync with the way of gitsync synchronise file... it's switch directory link... and fs notify don't detect each time the the update... in that way i used touch file as indicator (not the config file) of change (fsnotify watch this file in place of the configuration file), and i used the update hook of git sync to touch the file to trigger the reload.... ok we can used unix signal but it's not work on windows...
